src/Data/Functor/Foldable.hs

```diff
@@ -598,11 +603,37 @@ gchrono w m = ghylo (distGHisto w) (distGFutu m)
 
 -- | Mendler-style iteration
 mcata :: (forall y. (y -> c) -> f y -> c) -> Fix f -> c
-mcata psi = psi (mcata psi) . unFix
+mcata psi = c where c = psi c . unFix
+
+-- | Mendler-style version of 'para'
```

Comment:
either change all the other haddocks, or change this to *Mendler-style recursion*.

Also add `@since` annotation. (I'm optimistic that we can make `5.2.1` with just these changes).

I have no idea what is `zygo`. Does anyone uses `zygo` in practice?

Reply:
Done.

> I'm optimistic that we can make `5.2.1` with just these changes).

Sure, let's do that.

> Does anyone uses `zygo` in practice?

This is where it would be handy to have someone on the team who uses the library in practice! I think you're the one among us who uses it the most.

> I have no idea what is `zygo`

`zygo f (\(... (b,a) ...) -> ... b ...)` is a faster version of `para (\(... (t,a) ...) -> ... (cata f t) ...)`. This is the recursion scheme equivalent of writing a `go` helper which keeps more information than what your top-level function returns, in order to make sure you have all the information you need to perform a single step of the recursion without having to do O(n) work at each node. For example, suppose you want to find the deepest node in a tree. Here is a naïve way to do it:

```haskell
data Tree a
  = Leaf
  | Branch (Tree a) a (Tree a)
TH.makeBaseFunctor ''Tree

depth :: Tree a -> Int
depth = cata depthF

depthF :: Base (Tree a) Int -> Int
depthF = \case
  LeafF
    -> 0
  BranchF l _ r
    -> 1 + max l r

deepestValueA :: Tree a -> Maybe a
deepestValueA = \case
  Leaf
    -> Nothing
  Branch l x r
    | depth l > depth r
      -> deepestValueA l
    | depth l < depth r
      -> deepestValueA r
    | otherwise
      -> deepestValueA l <|> deepestValueA r <|> pure x
```

This is quite inefficient: we traverse all n nodes of the tree, and at each of them, we make a few O(n) calls to `depth`, thus leading to quadratic behaviour. The obvious solution is to keep track of the depth as we go up the tree instead of recomputing it for each node:

```haskell
deepestValueB :: Tree a -> Maybe a
deepestValueB = snd . go
  where
    go :: Tree a -> (Int, Maybe a)
    go = \case
      Leaf
        -> (0, Nothing)
      Branch l x r
        -> let (depthL, deepestValueL) = go l
               (depthR, deepestValueR) = go r
               deepestValue
                 | depthL > depthR
                   = deepestValueL
                 | depthL < depthR
                   = deepestValueR
                 | otherwise
                   = deepestValueL <|> deepestValueR <|> pure x
            in (1 + max depthL depthR, deepestValue)
```

This is faster, but it duplicates the logic of computing the depth of a node from the depth of its children. Let's look at the recursion-schemes version. First, the inefficient way:

```haskell
deepestValueC :: Tree a -> Maybe a
deepestValueC = para $ \case
  LeafF
    -> Nothing
  BranchF (l,deepestValueL) x (r,deepestValueR)
    | depth l > depth r
      -> deepestValueL
    | depth l < depth r
      -> deepestValueR
    | otherwise
      -> deepestValueL <|> deepestValueR <|> pure x
```

Since `depth` happens to be a `cata`, we can rewrite this to a zygo in order to get the efficient version:

```haskell
deepestValueD :: Tree a -> Maybe a
deepestValueD = zygo depthF $ \case
  LeafF
    -> Nothing
  BranchF (depthL,deepestValueL) x (depthR,deepestValueR)
    | depthL > depthR
      -> deepestValueL
    | depthL < depthR
      -> deepestValueR
    | otherwise
      -> deepestValueL <|> deepestValueR <|> pure x
```

And now the resulting code is not only fast, it is also reusing the `depthF` logic instead of duplicating it!

[Here](https://gist.github.com/gelisam/f64934146a31194d71e2883c24f4c879#file-flamegraphdiff-hs-L98) is a real-world situation in which I used the `go` approach in order to improve performance.